File: iosApp/iosApp/Modules/Account ✅/AccountWireframeFactory.swift
// Created by web3d3v on 11/02/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3lib

// MARK: - AccountWireframeFactory

protocol AccountWireframeFactory {
    func make(
        _ parent: UIViewController?,
        context: AccountWireframeContext
    ) -> AccountWireframe
}

// MARK: - DefaultAccountWireframeFactory

final class DefaultAccountWireframeFactory {
    private let currencyReceiveWireframeFactory: CurrencyReceiveWireframeFactory
    private let currencySendWireframeFactory: CurrencyCurrencyWireframeFactory
    private let currencySwapWireframeFactory: CurrencySwapWireframeFactory
    private let deepLinkHandler: DeepLinkHandler
    private let networksService: NetworksService
    private let currencyStoreService: CurrencyStoreService
    private let walletService: WalletService
    private let transactionService: IosEtherscanService

    init(
        currencyReceiveWireframeFactory: CurrencyReceiveWireframeFactory,
        currencySendWireframeFactory: CurrencyCurrencyWireframeFactory,
        currencySwapWireframeFactory: CurrencySwapWireframeFactory,
        deepLinkHandler: DeepLinkHandler,
        networksService: NetworksService,
        currencyStoreService: CurrencyStoreService,
        walletService: WalletService,
        transactionService: IosEtherscanService
    ) {
        self.currencyReceiveWireframeFactory = currencyReceiveWireframeFactory
        self.currencySendWireframeFactory = currencySendWireframeFactory
        self.currencySwapWireframeFactory = currencySwapWireframeFactory
        self.deepLinkHandler = deepLinkHandler
        self.networksService = networksService
        self.currencyStoreService = currencyStoreService
        self.walletService = walletService
        self.transactionService = transactionService
    }
}

extension DefaultAccountWireframeFactory: AccountWireframeFactory {

    func make(
        _ parent: UIViewController?,
        context: AccountWireframeContext
    ) -> AccountWireframe {
        DefaultAccountWireframe(
            parent,
            context: context,
            currencyReceiveWireframeFactory: currencyReceiveWireframeFactory,
            currencySendWireframeFactory: currencySendWireframeFactory,
            currencySwapWireframeFactory: currencySwapWireframeFactory,
            deepLinkHandler: deepLinkHandler,
            networksService: networksService,
            currencyStoreService: currencyStoreService,
            walletService: walletService,
            transactionService: transactionService
        )
    }
}

// MARK: - Assembler

final class AccountWireframeFactoryAssembler: AssemblerComponent {

    func register(to registry: AssemblerRegistry) {
        registry.register(scope: .instance) { resolver -> AccountWireframeFactory in
            DefaultAccountWireframeFactory(
                currencyReceiveWireframeFactory: resolver.resolve(),
                currencySendWireframeFactory: resolver.resolve(),
                currencySwapWireframeFactory: resolver.resolve(),
                deepLinkHandler: resolver.resolve(),
                networksService: resolver.resolve(),
                currencyStoreService: resolver.resolve(),
                walletService: resolver.resolve(),
                transactionService: resolver.resolve()
            )
        }
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/compose/account/AccountWireframeFactory.kt
package com.sonsofcrypto.web3wallet.android.modules.compose.account

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.services.currencyStore.CurrencyStoreService
import com.sonsofcrypto.web3lib.services.wallet.WalletService
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.common.AssemblerComponent
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistry
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistryScope
import com.sonsofcrypto.web3wallet.android.modules.compose.currencyreceive.CurrencyReceiveWireframeFactory
import com.sonsofcrypto.web3wallet.android.modules.compose.currencysend.CurrencySendWireframeFactory
import com.sonsofcrypto.web3wallet.android.modules.compose.currencyswap.CurrencySwapWireframeFactory
import com.sonsofcrypto.web3walletcore.modules.account.AccountWireframe
import com.sonsofcrypto.web3walletcore.modules.account.AccountWireframeContext
import com.sonsofcrypto.web3walletcore.services.etherScan.EtherScanService
import smartadapter.internal.extension.name

interface AccountWireframeFactory {
    fun make(parent: Fragment?, context: AccountWireframeContext): AccountWireframe
}

class DefaultAccountWireframeFactory(
    private val currencyReceiveWireframeFactory: CurrencyReceiveWireframeFactory,
    private val currencySendWireframeFactory: CurrencySendWireframeFactory,
    private val currencySwapWireframeFactory: CurrencySwapWireframeFactory,
    private val currencyStoreService: CurrencyStoreService,
    private val walletService: WalletService,
    private val etherScanService: EtherScanService,
): AccountWireframeFactory {

    override fun make(
        parent: Fragment?, context: AccountWireframeContext
    ): AccountWireframe = DefaultAccountWireframe(
        parent?.let { WeakRef(it) },
        context,
        currencyReceiveWireframeFactory,
        currencySendWireframeFactory,
        currencySwapWireframeFactory,
        currencyStoreService,
        walletService,
        etherScanService,
    )
}

class AccountWireframeFactoryAssembler: AssemblerComponent {

    override fun register(to: AssemblerRegistry) {

        to.register(AccountWireframeFactory::class.name, AssemblerRegistryScope.INSTANCE) {
            DefaultAccountWireframeFactory(
                it.resolve(CurrencyReceiveWireframeFactory::class.name),
                it.resolve(CurrencySendWireframeFactory::class.name),
                it.resolve(CurrencySwapWireframeFactory::class.name),
                it.resolve(CurrencyStoreService::class.name),
                it.resolve(WalletService::class.name),
                it.resolve(EtherScanService::class.name),
            )
        }
    }
}